Source/WebCore/css/StyleProperties.cpp

```diff
@@ -984,17 +984,89 @@ String StyleProperties::getGridTemplateValue() const
     return result.toString();
 }
 
+static bool gridAutoFlowContains(const RefPtr<CSSValue>& autoFlow, CSSValueID id)
+{
+    if (is<CSSValueList>(autoFlow)) {
+        for (auto& currentValue : downcast<CSSValueList>(*autoFlow)) {
+            if (isValueID(currentValue, id))
+                return true;
+        }
+        return false;
+    }
+    return isValueID(autoFlow, id);
+}
+
 String StyleProperties::getGridValue() const
 {
     // If none of the implicit track properties have been set, then
     // this shorthand can be represented using the grid-template shorthand
     // format.
-    if (isValueID(getPropertyCSSValue(CSSPropertyGridAutoColumns), CSSValueAuto)
-        && isValueID(getPropertyCSSValue(CSSPropertyGridAutoRows), CSSValueAuto)
-        && isValueID(getPropertyCSSValue(CSSPropertyGridAutoFlow), CSSValueRow)) {
+    auto autoColumns = getPropertyCSSValue(CSSPropertyGridAutoColumns);
+    auto autoRows = getPropertyCSSValue(CSSPropertyGridAutoRows);
+    auto autoFlow = getPropertyCSSValue(CSSPropertyGridAutoFlow);
+    if (isValueID(autoColumns, CSSValueAuto)
+        && isValueID(autoRows, CSSValueAuto)
+        && isValueID(autoFlow, CSSValueRow)) {
         return getGridTemplateValue();
     }
-    return getGridShorthandValue(gridShorthand());
+
+    if (!isValueID(getPropertyCSSValue(CSSPropertyGridTemplateAreas), CSSValueNone))
+        return String();
+
+    auto rowsIndex = findPropertyIndex(CSSPropertyGridTemplateRows);
+    if (rowsIndex == -1)
+        return String();
+    auto columnsIndex = findPropertyIndex(CSSPropertyGridTemplateColumns);
+    if (columnsIndex == -1)
+        return String();
+
+    if (!autoColumns || !autoRows)
+        return String();
+
+    auto rows = propertyAt(rowsIndex);
+    auto columns = propertyAt(columnsIndex);
+
+    StringBuilder result;
+
+    if (gridAutoFlowContains(autoFlow, CSSValueColumn)) {
+        if (!isValueID(autoRows, CSSValueAuto) || !isValueID(columns.value(), CSSValueNone))
+            return String();
+
+        if (rows.isImplicit())
+            return String();
+
+        result.append(rows.value()->cssText());
+        result.append(" / auto-flow");
+        if (gridAutoFlowContains(autoFlow, CSSValueDense))
+            result.append(" dense");
+
+        if (!isValueID(autoColumns, CSSValueAuto)) {
+            result.append(" ");
+            result.append(autoColumns->cssText());
+        }
+
+        return result.toString();
+    }
+
```

Comment:
Please check that `grid-auto-flow` contains `row`. Otherwise:

```js
document.body.style.cssText = "grid: auto-flow 1px / 2px; grid-auto-flow: inherit";
document.body.style.grid; // "auto-flow 1px / 2px"
```

Reply:
Actually, `row` is optional and `dense` is equivalent to `row dense`. So check that it contains either `row` or `dense`, or check that is not a css-wide value, or something like that.

Reply:
`grid` seems like a rather weird shorthand property which doesn't roundtrip properly.

This example (`grid: auto-flow 1px / 2px;`) is almost identical to `Example 27` in the spec, where it says that this value for `grid` also implies `grid-auto-flow: row;`.

To serialize `auto-flow 1px / 2px` would imply that `grid-auto-flow` is `row`, but it's really `inherit`.

As far as I can tell, there is no valid serialisation for `grid` which would correctly display the current style state, which is why I return `""`. Gecko appears to be doing the same thing, and I assume blink too.

Possibly the spec should be more clear on this? It seems like the shorthand isn't really serialisable using normal CSS rules.

Reply:
Apologies, that's exactly what you were suggesting. I'll add tests :)

Reply:
Exactly, empty string when the shorthand can't represent all the values of the longhands: https://drafts.csswg.org/cssom/#serializing-css-values